### PR TITLE
WIP: Simple Cython implementation

### PR DESCRIPTION
**Pull request by matusvalo**
This PR is an alternative to #258.

It consists of 3 parts:

1. Use `__slots__` in GenericContent, since should be more lightweight than regular class
2. Removes and simplifies parameters of functions in serialization.py - this is prerequisite for cythonizing + this should be faster in cpython 3.7+ [1] [2]
3. simple cythonization of serialization.py hot paths using pure python mode - this enables two possible ways of using library a) cythonized b) not cythonized

The advantage of this approach is to have one single code base which is already well tested and tuned by users. The disadvantage is that not everything can be converted to "pure" C.

What is pending:
1. performance tests (should be faster but no tests are done)
2. There is one single unittests failing (all other are passed) - this needs to be decided if the test is ignored on cython or not
3. setup.py needs to be refactored to somehow allow user to choose if he wants use cythonized version or pure python
4. CI must be set to execute tests cythoned and pure python version of library
5. I have even more aggressive cythonization of serialization.py but it is partialy broken and needs some small adjustments of serialization.py

[1] https://bugs.python.org/issue26110
[2] https://docs.python.org/3.7/whatsnew/3.7.html#cpython-bytecode-changes

### CODE REVIEW

**Comment by matusvalo** [comment]
There is only one difference of behavior between cythonized and pure python version of library (as long as I know) - and this is causing the only one unittest failing:

cythonized version of library due translating python function to C function is not raising `FrameSyntaxError` on some places - e.g.:

https://github.com/celery/py-amqp/blob/ed759687c069b12c914d8ebdab26ef34d068fca9/amqp/serialization.py#L141-L144

In cython it is tranlated to:
```
+141:         raise FrameSyntaxError(
  /*else*/ {
    __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_FrameSyntaxError); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 141, __pyx_L1_error)
    __Pyx_GOTREF(__pyx_t_2);
+142:             'Unknown value in table: {0!r} ({1!r})'.format(
+143:                 ftype, type(ftype)))
+144:     return val, offset
```

But this should **not** happen when broker is correctly implemented - but in any case cythonized version should not be used with nonstandard/experimental brokers.

**Comment by matusvalo** [comment]
Here is cythonization report of serialization.py:

[serialization.zip](https://github.com/celery/py-amqp/files/2912808/serialization.zip)

File: amqp/channel.py [hunk 1]
```diff
@@ -1600,13 +1600,11 @@ def basic_consume(self, queue='', consumer_tag='', no_local=False,
...
                           exchange, routing_key, msg):
         msg.channel = self
-        msg.delivery_info = {
-            'consumer_tag': consumer_tag,
-            'delivery_tag': delivery_tag,
-            'redelivered': redelivered,
-            'exchange': exchange,
-            'routing_key': routing_key,
-        }
+        msg.delivery_info['consumer_tag'] = consumer_tag
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply amqp/channel.py]: See comment below.

File: amqp/channel.py [hunk 2]
```diff
@@ -1670,13 +1668,11 @@ def _on_get_empty(self, cluster_id=None):
...
                    message_count, msg):
         msg.channel = self
-        msg.delivery_info = {
-            'delivery_tag': delivery_tag,
-            'redelivered': redelivered,
-            'exchange': exchange,
-            'routing_key': routing_key,
-            'message_count': message_count
-        }
+        msg.delivery_info['delivery_tag'] = delivery_tag
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply amqp/channel.py]: I am not sure why I did it. Maybe it was something forgotten when I did experiments. I have reverted it for now.

**Comment by thedrow** [comment]
I still think we should avoid C extensions and unsafe code.
Cython code cannot be shared with CFFI code. The Rust code can be.

**Comment by auvipy** [comment]
then we could try this in another fork. celery should be python3 only

**Comment by thedrow** [comment]
I don't think you understand what I mean.
C extensions are not the right way to go.
Rust is when it is optional.

**Comment by auvipy** [comment]
https://twitter.com/pwang/status/1203336880886075392/photo/1

**Comment by matusvalo** [comment]
Also aiohttp is using cython - see:

https://github.com/aio-libs/aiohttp/tree/master/aiohttp

**Comment by matusvalo** [comment]
Closed in favor of #311

**Pull request closed without merging**